subsys/bluetooth/host/conn.c

```diff
@@ -2685,7 +2685,7 @@ int bt_conn_auth_pairing_confirm(struct bt_conn *conn)
 
 uint8_t bt_conn_index(struct bt_conn *conn)
 {
-	uint8_t index;
+	uintptr_t index;
```

Comment:
The strictly correct type for this would be `ptrdiff_t` as it's used for the difference between pointers (and is signed).

It appears that the result of subtraction is always non-negative, so the corresponding unsigned type is `size_t`.

`uintptr_t` is usually used for representing pointer values, not pointer differences.

Reply:
That's a very good point. 

Since we are subtracting two values, without checking if the value is going to be positive or negative, I guess the correct solution would require a signed variable. Alternative is to also add an __ASSERT if the pointer checked is lower than the pool that it's checked against.

Reply:
We generally don't bother sign-checking this sort of thing, since it's "known" that the pointer is a member of the sequence starting at the base pointer that's subtracted.  For MISRA reasons it should be `index = (size_t)(ptr - base);` but that's not generally done.

Reply:
> We generally don't bother sign-checking this sort of thing, since it's "known" that the pointer is a member of the sequence starting at the base pointer that's subtracted. 

If we know that the pointer is a member of the sequence, why bother checking the index at all? I thought the point of the check was to check if the pointer indeed was a member of the sequence. 

What value does the `size_t` typecast add here?

Reply:
OK, I was again misled because the deltas displayed don't cover enough of what the code does, so I didn't notice the asserts.

What problem are you trying to solve here?  The problem description is not detailed.

Presumably at some point somebody passed in a `conn` that was either **not** a member of the corresponding `foo_conns` array.

Was it a null pointer?  Was it a pointer to a valid `conn` but the `conn->type` incorrectly identified the sequence it was to belong to?  Or is it that somebody set `CONFIG_BT_MAX_FOO_CONN` greater than 127 so the truncation discarded important data?

Clearly `conn` can't be null, because then the switch argument would cause a fault.

I have to guess that what happened was either the `type` was wrong or there were more than 127 connections of the given type.  In any case, you need to check that the pointer is strictly within the range of the connection array, which means you have to check for negative values.

In short, if you want to do this correctly, you should do:

```
    ptrdiff_t index;
    ...
    index = conn - foo_conns;
    __ASSERT(index >=0 && index < CONFIG_BT_MAX_FOO_CONNS);
    ...
    return (uint8_t)index;
    ```
```

If you want to obscurely optimize away the check for negative values instead of letting the compiler do that, then do:

```

    size_t index;
    ...
    index = (size_t)(conn - foo_conns); // negatives become large positives
    __ASSERT(index < CONFIG_BT_MAX_FOO_CONNS);
    ...
    return (uint8_t)index;
    ```
```

The cast is required by MISRA because without it you'd be assigning a signed potentially negative value into a variable that can only hold an unsigned value, and you should be explicit that the value conversion is intentional.

Reply:
I think @pabigot proposed solution would indeed be the "correct" way of doing this. 

```
ptrdiff_t index;
...
index = conn - foo_conns;
__ASSERT(index >= 0 && index < CONFIG_BT_MAX_FOO_CONNS);
...
return (uint8_t)index;
```

Reply:
@pabigot I was trying to fix the assert, which had the problem that it truncated before the assert, which could lead to passing the assert with a pointer that did not point to the foo_conns array.

The purpose of the assert is to check that the pointer is part of the sequence. I have updated to explicit check, and will let the compiler do the optimization instead of the obscure optimization.

Thank you for the detailed feedback.